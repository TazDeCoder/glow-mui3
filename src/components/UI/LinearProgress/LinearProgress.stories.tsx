import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { LinearProgress } from './LinearProgress';
import { Flex } from '../../Utils';

export default {
  title: 'Components/UI/LinearProgress',
  component: LinearProgress,
} as ComponentMeta<typeof LinearProgress>;

export const Default: ComponentStory<typeof LinearProgress> = (args) => (
  <LinearProgress sx={{ width: 320 }} label="Loading..." {...args} />
);

export const Colors: ComponentStory<typeof LinearProgress> = (args) => (
  <Flex width={320} flexDirection="column" gap={2}>
    <LinearProgress {...args} color="primary" label="Primary" />
    <LinearProgress {...args} color="secondary" label="Secondary" />
    <LinearProgress {...args} color="info" label="Info" />
    <LinearProgress {...args} color="error" label="Error" />
    <LinearProgress {...args} color="warning" label="Warning" />
    <LinearProgress {...args} color="success" label="Success" />
  </Flex>
);

export const Variants: ComponentStory<typeof LinearProgress> = (args) => (
  <Flex width={320} flexDirection="column" gap={2}>
    <LinearProgress {...args} variant="indeterminate" label="Indeterminate" />
    <LinearProgress {...args} variant="determinate" label="Determinate" />
  </Flex>
);
