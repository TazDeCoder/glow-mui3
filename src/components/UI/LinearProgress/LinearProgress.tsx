/** @jsxRuntime classic */
/** @jsx jsx */

import { forwardRef } from 'react';
import { keyframes } from '@emotion/react';
import { jsx, ThemeUICSSObject, ThemeUIStyleObject } from '@theme-ui/core';

import { Box, BoxOwnProps, BoxProps, Flex } from '../../Utils';
import { Typography } from '../Typography';
import { __internalProps } from '../../shared/utils';
import type { Assign, ForwardRef, ColorTheme } from '../../types';

export interface LinearProgressProps
  extends Assign<React.ComponentPropsWithRef<'progress'>, BoxOwnProps> {
  /** Color theme used for the component
   * @default 'primary'
   */
  color?: ColorTheme;
  /** Variant style of the component
   * @default 'indeterminate'
   */
  variant?: 'indeterminate' | 'determinate';
  /** Value of the progress indicator */
  value?: number;
  /** Label associated to the component */
  label?: string;
}

const indeterminate1Keyframe = keyframes({
  '0%': {
    width: '100%',
    left: '-100%',
  },
  '60%': {
    width: '0%',
    left: '100%',
  },
  '100%': {
    width: '0%',
    left: '100%',
  },
});

// TODO: Add follow-up keyframe of first
// const indeterminate2Keyframe = keyframes({});

const __progressCss: ThemeUICSSObject = {
  position: 'absolute',
  bottom: 0,
  left: 0,
  transformOrigin: 'left',
  width: '100%',
  border: 'none',
  transition: 'standard',
};

export const LinearProgress: ForwardRef<HTMLDivElement, LinearProgressProps> =
  forwardRef(
    (
      {
        color = 'primary',
        variant = 'indeterminate',
        value,
        label,
        sx,
        ...props
      },
      ref,
    ) => {
      const style: ThemeUIStyleObject = {
        ...(variant === 'determinate' && {
          width: `${value}%`,
          transition: 'emphasized',
        }),
        '&::-webkit-progress-bar': {
          bg: `${color}.main`,
        },
        '&::-moz-progress-bar': {
          bg: `${color}.main`,
        },
      };

      return (
        <Flex flexDirection="column" gap={2} sx={sx}>
          <Box
            {...__internalProps({
              __css: {
                position: 'relative',
                height: 4,
                overflow: 'hidden',
                bg: 'surface.variant',
              },
            })}
          >
            {variant === 'indeterminate' ? (
              <Box
                ref={ref}
                as="progress"
                aria-label="progress-indeterminate"
                sx={{
                  ...style,
                  animation: `${indeterminate1Keyframe} 3.1s cubic-bezier(0.65, 0.815, 0.735, 0.395) infinite`,
                }}
                {...(props as BoxProps)}
                {...__internalProps({ __css: __progressCss })}
              />
            ) : (
              <Box
                ref={ref}
                as="progress"
                aria-label="progress-determinate"
                sx={style}
                {...(props as BoxProps)}
                {...__internalProps({ __css: __progressCss })}
              />
            )}
          </Box>
          {label && (
            <Typography variant="label" color={`${color}.main`}>
              {label}
            </Typography>
          )}
        </Flex>
      );
    },
  );
