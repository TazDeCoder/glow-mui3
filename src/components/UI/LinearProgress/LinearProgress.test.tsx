import { render } from '@testing-library/react';

import { LinearProgress } from './LinearProgress';

describe('<LinearProgress />', () => {
  describe('Snapshots', () => {
    it('renders correctly', () => {
      const { container } = render(<LinearProgress />);
      expect(container).toMatchSnapshot();
    });
  });
});
