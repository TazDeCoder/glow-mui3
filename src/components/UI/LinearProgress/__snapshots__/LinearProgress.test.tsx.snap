// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinearProgress /> Snapshots renders correctly 1`] = `
@keyframes animation-0 {
  0% {
    width: 100%;
    left: -100%;
  }

  60% {
    width: 0%;
    left: 100%;
  }

  100% {
    width: 0%;
    left: 100%;
  }
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 8px;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  height: 4px;
  overflow: hidden;
  background-color: surface.variant;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: absolute;
  bottom: 0;
  left: 0;
  transform-origin: left;
  width: 100%;
  border: none;
  -webkit-transition: standard;
  transition: standard;
  -webkit-animation: animation-0 3.1s cubic-bezier(0.65, 0.815, 0.735, 0.395) infinite;
  animation: animation-0 3.1s cubic-bezier(0.65, 0.815, 0.735, 0.395) infinite;
}

.emotion-2::-webkit-progress-bar {
  background-color: primary.main;
}

.emotion-2::-moz-progress-bar {
  background-color: primary.main;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <progress
        aria-label="progress-indeterminate"
        class="emotion-2"
      />
    </div>
  </div>
</div>
`;
